modules/imgcodecs/src/grfmt_jpeg2000_openjpeg.cpp

```diff
@@ -0,0 +1,694 @@
+            copyToMat(ImageComponents { inImg.comps[2].data, inImg.comps[1].data, inImg.comps[0].data },
+                      tmp, shift);
+            cvtColor(tmp, outImg, COLOR_BGR2GRAY);
+        }
+        return true;
+    }
+
+    if (inChannels >= 3)
+    {
+        // Assume RGB (+ alpha) for 3 channels -> BGR
+        ImageComponents incomps { inImg.comps[2].data, inImg.comps[1].data, inImg.comps[0].data };
+        // Assume RGBA for 4 channels -> BGRA
+        if (outChannels > 3)
+        {
+            incomps.push_back(inImg.comps[3].data);
+        }
+        copyToMat(std::move(incomps), outImg, shift);
+        return true;
+    }
+    CV_LOG_ERROR(nullptr,
+                 cv::format("OpenJPEG2000: unsupported conversion from %d components to %d for SRGB image decoding",
+                            inChannels, outChannels));
+    return false;
+}
+
+bool decodeGrayscaleData(const opj_image_t& inImg, cv::Mat& outImg, uint8_t shift)
+{
+    using ImageComponents = std::vector<const OPJ_INT32*>;
+
+    const int inChannels = inImg.numcomps;
+    const int outChannels = outImg.channels();
+
+    if (outChannels == 1 || outChannels == 3)
+    {
+        copyToMat(ImageComponents(outChannels, inImg.comps[0].data), outImg, shift);
+        return true;
+    }
+    CV_LOG_ERROR(nullptr,
+                 cv::format("OpenJPEG2000: unsupported conversion from %d components to %d for Grayscale image decoding",
+                            inChannels, outChannels));
+    return false;
+}
+
+bool decodeSYCCData(const opj_image_t& inImg, cv::Mat& outImg, uint8_t shift)
+{
+    using ImageComponents = std::vector<const OPJ_INT32*>;
+
+    const int inChannels = inImg.numcomps;
+    const int outChannels = outImg.channels();
+
+    if (outChannels == 1) {
+        copyToMat(ImageComponents { inImg.comps[0].data }, outImg, shift);
+        return true;
+    }
+
+    if (outChannels == 3 && inChannels >= 3) {
+        copyToMat(ImageComponents { inImg.comps[0].data, inImg.comps[1].data, inImg.comps[2].data },
+                  outImg, shift);
+        cvtColor(outImg, outImg, COLOR_YUV2BGR);
+        return true;
+    }
+
+    CV_LOG_ERROR(nullptr,
+                 cv::format("OpenJPEG2000: unsupported conversion from %d components to %d for YUV image decoding",
+                            inChannels, outChannels));
+    return false;
+}
+} // namespace <anonymous>
+
+/////////////////////// Jpeg2KOpjDecoder ///////////////////
+
+Jpeg2KOpjDecoder::Jpeg2KOpjDecoder()
+{
+    static const unsigned char signature[] = { 0, 0, 0, 0x0c, 'j', 'P', ' ', ' ', 13, 10, 0x87, 10 };
+    m_signature = String((const char*)(signature), sizeof(signature));
+    // FIXME: raw code stream
+}
+
+
+ImageDecoder Jpeg2KOpjDecoder::newDecoder() const
+{
+    return makePtr<Jpeg2KOpjDecoder>();
+}
+
+bool Jpeg2KOpjDecoder::readHeader()
+{
+    stream_.reset(opj_stream_create_default_file_stream(m_filename.c_str(), OPJ_STREAM_READ));
```

Comment:
> m_filename.c_str()

Looks like `imdecode()` from in-memory buffers is not supported at this moment.

Reply:
Added support, based on the following PR:
 * [PR 583](https://github.com/uclouvain/openjpeg/pull/583)
 * [PR 1212](https://github.com/uclouvain/openjpeg/pull/1212)